interop-testing/src/test/java/io/grpc/testing/integration/ProxyTest.java

```diff
@@ -0,0 +1,233 @@
+package io.grpc.testing.integration;
+
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Ignore;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.SocketImpl;
+import java.net.UnknownHostException;
+import java.util.concurrent.TimeUnit;
+
+@RunWith(JUnit4.class)
+public class ProxyTest {
+
+  private int serverPort = 5001;
+  private int proxyPort = 5050;
+  private String loopBack = "127.0.0.1";
+
+  // These all work individually right now but running them sequentially causes errors
+  // because shutDown() is not clean. Holding off on fixing it right now because the threading
+  // model for the proxy is probably going to change (which will hopefully make shutting down
+  // easier).
+
+  @Test
+  @Ignore
+  public void smallLatency() throws UnknownHostException, IOException, InterruptedException {
+    Server server = new Server();
+    (new Thread(server)).start();
+
+    int latency = (int) TimeUnit.MILLISECONDS.toNanos(10);
+    TrafficControlProxy p = new TrafficControlProxy(1024 * 1024, latency, TimeUnit.NANOSECONDS);
+    (new Thread(new Proxy(p))).start();
+    Thread.sleep(100);
+    Socket client = new Socket(loopBack, proxyPort);
+    DataOutputStream clientOut = new DataOutputStream(client.getOutputStream());
+    DataInputStream clientIn = new DataInputStream(client.getInputStream());
+    byte[] message = new byte[1];
+
+    // warm up
+    for (int i = 0; i < 10; i++) {
+      clientOut.write(message, 0, 1);
+      clientIn.read(message);
+    }
+    // test
+    long start = System.nanoTime();
+    clientOut.write(message, 0, 1);
+    clientIn.read(message);
+    long stop = System.nanoTime();
+
+    long rtt = (stop - start);
+    // NOTE: arbitrarily picked acceptable error to be 2 milliseconds
+    assertEquals(latency, rtt, TimeUnit.MILLISECONDS.toNanos(2));
+
+    p.shutDown();
+    server.shutDown();
+    client.close();
+  }
+
+  @Test
+  @Ignore
+  public void bigLatency() throws UnknownHostException, IOException, InterruptedException {
+    Server server = new Server();
+    (new Thread(server)).start();
```

Comment:
FYI threads can be started, but they can't be explicitly ended from the outside.   What you can do is wait for them to finish.  This is called joining.  You should keep a reference to the thread, and call thread.join() from this thread to make sure it finishes.